nengo/neurons.py

```diff
@@ -727,7 +814,155 @@ def step_math(self, dt, J, spiked, voltage, recovery):
+
+    Spikes at regular intervals based on the rates of the base neuron type.
+
+    Parameters
+    ----------
+    base_type : NeuronType
+        A rate-based neuron type to convert to a regularly spiking neuron.
+    amplitude : float
+        Scaling factor on the neuron output. Corresponds to the relative
+        amplitude of the output spikes of the neuron.
+    min_voltage : float
+        The firing threshold for negative spikes. Only has an effect if
+        ``base_type.negative``.
+    """
+
+    min_voltage = NumberParam("min_voltage", low=-1, high=0)
+
+    def __init__(self, base_type, amplitude=1.0, min_voltage=-0.1):
+        super().__init__(base_type, amplitude=amplitude)
+
+        if "voltage" in base_type.probeable:
+            raise ValidationError(
+                "Cannot already include `voltage` in `probable`",
+                attr="base_type",
+                obj=self,
+            )
+
+        self.min_voltage = min_voltage if self.negative else 0.0
+
+    @property
+    def probeable(self):
+        base = list(self.base_type.probeable)
+        base[base.index("rates")] = "spikes"
+        base.append("voltage")
+        return tuple(base)
+
+    def step_math(self, dt, J, output, voltage, *base_states):
+        # Sets output to the desired rates
+        self.base_type.step_math(dt, J, output, *base_states)
+
+        voltage += dt * output
+        if self.min_voltage == 0:
+            n_spikes = np.floor(voltage)
+        else:
+            # add positive spikes for any voltages over 1
+            n_spikes = np.maximum(np.floor(voltage), 0)
+            # add negative spikes for any voltages under min_voltage
+            n_spikes += np.minimum(np.floor(voltage - self.min_voltage), 0)
+
+        output[...] = (self.amplitude / dt) * n_spikes
+        voltage -= n_spikes
+
+
+class PoissonSpiking(_Spiking):
+    """Turn a rate neuron model into a spiking one with Poisson spiking statistics.
+
+    Spikes with Poisson probability based on the rates of the base neuron type.
+
+    Parameters
+    ----------
+    base_type : NeuronType
+        A rate-based neuron type to convert to a Poisson spiking neuron.
+    amplitude : float
+        Scaling factor on the neuron output. Corresponds to the relative
+        amplitude of the output spikes of the neuron.
+    """
+
+    def step_math(self, dt, J, output, rng, *base_states):
+        # Sets output to the desired rates
+        rates = np.zeros_like(output)
+        self.base_type.step_math(dt, J, rates, *base_states)
+
+        signs = None
+        if self.negative:
+            signs = np.sign(rates)
+            rates = np.abs(rates)
+
+        n_spikes = rng.poisson(rates * dt, output.size)
+        output[...] = (self.amplitude / dt) * (
+            n_spikes if signs is None else n_spikes * signs
+        )
+
+
+class StochasticSpiking(_Spiking):
+    """Turn a rate neuron model into a spiking one with semi-Poisson spiking statistics.
```

Comment:
Looks like semi-Poisson already has a meaning other than what we mean here (e.g., https://arxiv.org/abs/cond-mat/0511171).

Reply:
Hmm, maybe "Poisson-like"? I wanted to get across the fact that spikes are independently generated each timestep. Maybe we just say that right out: "one with independent spiking statistics"? This seems to me to be the main way to do that other than true Poisson. We could even rename the whole class to `IndependentSpiking`, since it might be a bit more descriptive than `StochasticSpiking`.

Reply:
Or what about `DitheredSpiking`, since this is actually dithering?

Reply:
Decided to keep `StochasticSpiking` and make a specific reference to "stochastic rounding" in the docstring.